# xylose_optimization_project

Updated 10-19-2021

Summary: 

##1) Gene families for XYL1, XYL2, XYL3 were pulled out of 332 assemblies using an ORF-Finder -> HMMER pipeline written by JW and executed on scarcity (code not part of repo). 
- All sequences were assigned KEGG orothology using BLASTkoala. Phylogenetic trees of gene families were constructed using FastTree with KEGG orthology overlaied. The monophyletic clade containing all XYL homologs was id'd and XYL sequences were pulled out for further analysis. 
- Abbe and Rishitha worked with sequences to eliminate dubious sequences and ensure quality start sites and accurate lengths. 
- IQtree ML phylogenies of protein sequence were built. 
- Analysis for duplication events?

##2) Growth rate data was obtained from Dana and compared to gene content.
-Qualitative growth data for species not assayed by D.O. were retrieved from Opulente et. al BMC Biology 2018.

##3) Codon optimization was assessed as in Labella et al. PLoS Bio 2021 using data from Labella et al. PLoS Gen 2019
-
