# xylose_optimization_project

## Part 1
**R programming fundamentals with swirl.**

1. Open Rstudio. 
2. Use the console to install swirl. 
`install.packages("swirl")`  
3. Load the packages. 
`library(swirl)`  
4. Follow the prompts to complete the *R Programming - 1. Basic Building Blocks* course. 
5. After completing lesson 1, work through lessons R programming lessons 2 through 7. 

## Part 2
**ortholog trees**  
*I created new protein alignments in data/alnmnts/pep_ortholog_alnmnts - use these for below.*
1. Create maximum liklihood trees for the protein alignments in the xylose_optimization_project/data/alnmnts/pep_ortholog_alnmnts directory by uploading the fasta files to https://raxml-ng.vital-it.ch/#/. Be sure to reset Datatype to Protein/AA.  
2. Finish writing your `for` loops to populate the ortholog binary matrix. I swapped out the tkl1_cds_MSA.tsv file for a new one that includes more orthologs. *Be sure to re-run all your code with the new, updated files!!* 

